java/src/main/java/actorworkflow/CustomerLoyaltyWorkflowImpl.java

```diff
@@ -0,0 +1,147 @@
+            logger.info("Promoting customer!");
+            customer = customer.withStatus(tierToPromoteTo);
+            activities.sendEmail(EmailStrings.EMAIL_PROMOTED.formatted(tierToPromoteTo.name()));
+        }
+    }
+
+    @Override
+    public void inviteGuest(Customer guest) {
+        logger.info("Checking to see if customer can invite guests.");
+        if (Customer.canAddGuest(customer)) {
+            logger.info("Customer is allowed to invite guests; attempting to start workflow for guest ID {}.",
+                    guest.customerId());
+            customer.guests().add(guest);
+
+            StatusTier guestMinStatus = StatusTier.previous(customer.status());
+            guest = guest.withStatus(guestMinStatus);
+
+            String guestWorkflowId = Shared.WORKFLOW_ID_FORMAT.formatted(guest.customerId());
+            ChildWorkflowOptions options =
+                    ChildWorkflowOptions.newBuilder()
+                            .setWorkflowId(guestWorkflowId)
+                            .setParentClosePolicy(ParentClosePolicy.PARENT_CLOSE_POLICY_ABANDON)
+                            .build();
+
+            CustomerLoyaltyWorkflow child = Workflow.newChildWorkflowStub(CustomerLoyaltyWorkflow.class, options);
+
+            boolean alreadyStarted = false;
+            try {
+                Promise<WorkflowExecution> childExecution = Workflow.getWorkflowExecution(child);
+                Async.procedure(child::customerLoyalty, guest);
+
+                // Wait for child to start
+                childExecution.get();
+            } catch (WorkflowExecutionAlreadyStarted e) {
+                logger.info("Guest customer workflow already started and is a direct child.");
+                alreadyStarted = true;
+            } catch (Exception e) {
+                if (e.getCause() instanceof WorkflowExecutionAlreadyStarted) {
+                    logger.info("Guest customer workflow already started.");
+                } else {
+                    throw e;
+                }
+                alreadyStarted = true;
+            }
+
+            if (alreadyStarted) {
+                // Reset child to ensure we're actually working with the latest running execution
+                child = Workflow.newExternalWorkflowStub(CustomerLoyaltyWorkflow.class, guestWorkflowId);
+
+                logger.info("Signaling to ensure that they're at least \"{}\" status.", guestMinStatus.name());
+                child.ensureMinimumStatus(guestMinStatus);
+
+                activities.sendEmail(EmailStrings.EMAIL_GUEST_MIN_STATUS.formatted(guestMinStatus.name()));
+            } else {
+                activities.sendEmail(EmailStrings.EMAIL_GUEST_INVITED);
+            }
+        }
+    }
+
+    @Override
+    public void ensureMinimumStatus(StatusTier status) {
+        logger.info("Ensuring that status is at minimum {}.", status.name());
+        while (customer.status().minimumPoints() < status.minimumPoints()) {
```

Comment:
dont think blocking inside signal method is recommended. you can delay processing of other signals

Reply:
I don't think this really counts as blocking since the loop is bounded by the length of `StatusTier.STATUS_TIERS`.

Regardless, in the other languages, I rewrote the status handling to be easier to read, which I may do here (particularly if I take the time to do your suggestion of moving status handling to activities).